### PR TITLE
Provide onset of symptoms before upload (EXPOSUREAPP-1605)

### PR DESCRIPTION
**Pull request by Oliver-Zimmerman**
### Description

Implemented feature where the user provides the onset of symptoms before uploading a positive test result. 
- Created custom calendar widget
- Replaced generation of transmission risk vector
- Created fragments for symptom introduction and symptom calendar screens
- Created necessary XML views for fragments 
- Adjusted SubmissionViewModel for required fragments

### Steps to reproduce

1. Submit a new positive test via QR code. 
2. Select one of three buttons on symptom introduction screen, once one button is selected, the others should be unselected. Also, the user should not be able to navigate until a button has been selected. 
3. Navigate to calendar screen, select either an exact date or range via buttons. Selecting a date or range should unselect other options. User should not be able to navigate until an option is selected.

### CODE REVIEW

File: Corona-Warn-App/src/main/res/values/styles.xml [hunk 1]
```diff
@@ -140,6 +140,35 @@
...
+        <item name="android:background">@drawable/card</item>
+        <item name="fontFamily">sans-serif-medium</item>
+        <item name="android:fontFamily">sans-serif-medium</item>
+        <item name="android:textStyle">normal</item>
+        <item name="android:textAllCaps">false</item>
+        <item name="android:textSize">20sp</item>
+        <item name="android:letterSpacing">0.0125</item>
+    </style>
+
+    <style name="greenButtonNotPressed">
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/styles.xml:155]: isn't used

File: Corona-Warn-App/src/main/res/values/styles.xml [hunk 2]
```diff
@@ -140,6 +140,35 @@
...
+        <item name="android:padding">@dimen/card_padding</item>
+        <item name="android:background">@drawable/card</item>
+        <item name="android:backgroundTint">@color/colorSurface2</item>
+    </style>
+
+    <style name="greenButtonPressed">
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/styles.xml:161]: isn't used

File: Corona-Warn-App/src/main/res/values-night/colors.xml [hunk 3]
```diff
@@ -53,4 +53,11 @@
     <color name="colorStableHairlineLight">#33FFFFFF</color>
     <color name="colorStableHairlineDark">#3317191A</color>
 
+    <!-- Calendar -->
+    <color name="colorCalendarSelectedDayBackground">#5D6F80</color>
+    <color name="colorCalendarTodayBorder">#007FAD</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-night/colors.xml:58]: color already exists, you could remap it or reuse it.

File: Corona-Warn-App/src/main/res/values/colors.xml [hunk 4]
```diff
@@ -24,6 +24,8 @@
     <color name="colorTextSemanticGreen">#2E854B</color>
     <color name="colorTextSemanticNeutral">#5D6E80</color>
     <color name="colorTextTint">#007FAD</color>
+    <color name="colorTextGreenButtonPressed">#FFFFFF</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/colors.xml:27]: isn't used

File: Corona-Warn-App/src/main/res/values/colors.xml [hunk 5]
```diff
@@ -24,6 +24,8 @@
...
+    <color name="colorTextGreenButtonNotPressed">#000000</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/colors.xml:28]: isn't used

File: Corona-Warn-App/src/main/res/values/colors.xml [hunk 6]
```diff
@@ -52,5 +54,14 @@
     <color name="colorStableLight">#FFFFFF</color>
     <color name="colorStableHairlineLight">#33FFFFFF</color>
     <color name="colorStableHairlineDark">#3317191A</color>
+    <color name="colorGreenButtonPressed">#5D6E80</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/colors.xml:57]: isn't used

File: Corona-Warn-App/src/main/res/values/colors.xml [hunk 7]
```diff
@@ -52,5 +54,14 @@
...
+    <color name="colorGreenButtonNotPressed">#F5F5F5</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/colors.xml:58]: isn't used

File: Corona-Warn-App/src/main/res/values-night/colors.xml [hunk 8]
```diff
@@ -53,4 +53,11 @@
     <color name="colorStableHairlineLight">#33FFFFFF</color>
     <color name="colorStableHairlineDark">#3317191A</color>
 
+    <!-- Calendar -->
+    <color name="colorCalendarSelectedDayBackground">#5D6F80</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-night/colors.xml:57]: don't has be overwritten if it exists in the default colors.

File: Corona-Warn-App/src/main/res/values/colors.xml [hunk 9]
```diff
@@ -52,5 +54,14 @@
...
 
+    <!-- Calendar -->
+    <color name="colorCalendarSelectedDayBackground">#5D6F80</color>
+    <color name="colorCalendarTodayBorder">#007FAD</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/colors.xml:62]: color already exists

File: Corona-Warn-App/src/main/res/values/colors.xml [hunk 10]
```diff
@@ -52,5 +54,14 @@
...
+    <color name="colorCalendarTodayText">#007FAD</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values/colors.xml:63]: color already exists

File: Corona-Warn-App/src/main/res/values-night/colors.xml [hunk 11]
```diff
@@ -53,4 +53,11 @@
...
+    <color name="colorCalendarTodayBorder">#007FAD</color>
+    <color name="colorCalendarTodayText">#007FAD</color>
+    <color name="colorCalendarMonthText">#DEFFFFFF</color>
+    <color name="colorCalendarLayoutFocusOn">#FF5D6F80</color>
```
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-night/colors.xml:61]: don't has be overwritten if it exists in the default colors.

**Review by harambasicluka — CHANGES_REQUESTED** [review]
Only checked strings, colors & styles.

**Review by janetback — CHANGES_REQUESTED** [review]
UA review - could you please change the following. Thanks.
- under 'submission_symptom_symptom_bullets', please put a capital at the beginning of this string:  Allgemeine Schwäche und Abgeschlagenheit
- submission_symptom_calendar_body: an 'e' is missing at end of  genau: 
Selektieren Sie entweder das genaue Datum in dem Kalender oder wenn Sie sich nicht genau erinnern, eine der anderen Optionen.
- submission_symptom_less_seven: please put 'Tagen' in capital letters:
In den letzten 7 Tagen

**Review by janetback — DISMISSED** [review]
UA reviewed

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculationTest.kt [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Test
+    fun calculateSameYearSameMonth() {
+        var input = "27.08.2020"
+        val dateTime =
+            DateTime.parse(input, DateTimeFormat.forPattern(pattern))
+        val dates = CalendarCalculation().getDates(dateTime)
+
+        // First day - 3 of August
+        TestCase.assertEquals(3, dates.first().date.dayOfMonth)
```
Comment (jakobmoellersap) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculationTest.kt:20]: why do we use TestCase here?

**Review by jakobmoellerdev — CHANGES_REQUESTED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculation.kt [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+     * | -2| -1| 9 | +1| +2| +3| +4| <- Current Week (4th row)
+     * Code: (CurrentDayOfTheWeek * -1) + dayId
+     *
+     * Step 2: Apply week shift
+     * |_M_|_T_|_W_|_T_|_F_|_S_|_S_|
+     * | -9| -8| -7| -6| -5| -4| -3| <- Previous Week (3d row)
+     * | -2| -1| 9 | +1| +2| +3| +4| <- Current Week (4th row)
+     * Code: (DaysInWeekCount * (TotalWeeks - weekId)) * -1
+     */
+    fun getDates(currentDate: DateTime = DateTime(Instant.now())): List<CalendarAdapter.Day> {
```
Comment (jakobmoellersap) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculation.kt:105]: Do we need UTC here due to the keys being submitted to a server using UTC?

File: .idea/codeStyles/Project.xml [hunk 14]
```diff
@@ -1,9 +1,18 @@
 <component name="ProjectCodeStyleConfiguration">
   <code_scheme name="Project" version="173">
-     <JetCodeStyleSettings>
+    <JetCodeStyleSettings>
       <option name="PACKAGES_TO_USE_STAR_IMPORTS">
         <value>
-          <package name="kotlinx.android.synthetic" withSubpackages="true" static="false" />
+          <package name="kotlinx.android.synthetic" alias="false" withSubpackages="true" />
```
Comment (d4rken) [inline .idea/codeStyles/Project.xml:6]: Revert this please, it's a bug in Android Studio 4.0.1.

Comment (d4rken) [reply .idea/codeStyles/Project.xml:6]: I mean the whole change set to this file.

File: Corona-Warn-App/config/detekt.yml [hunk 15]
```diff
@@ -100,7 +100,7 @@ complexity:
     ignoreStringsRegex: '$^'
   TooManyFunctions:
     active: true
-    excludes: ['**/test/**', '**/androidTest/**', '**/*.Test.kt', '**/*.Spec.kt', '**/*.Spek.kt', '**/LocalData.kt', '**/formatter/*Helper.kt']
+    excludes: ['**/test/**', '**/androidTest/**', '**/*.Test.kt', '**/*.Spec.kt', '**/*.Spek.kt', '**/LocalData.kt', '**/formatter/*Helper.kt', '**/viewmodel/*']
```
Comment (d4rken) [inline Corona-Warn-App/config/detekt.yml:103]: Why is this necessary?

File: Corona-Warn-App/config/detekt.yml [hunk 16]
```diff
@@ -522,7 +522,8 @@ style:
       '**/*.Spek.kt',
       '**/CwaWebException.kt',
       '**/HttpErrorParser.kt',
-      '**/ErrorCodes.kt'
+      '**/ErrorCodes.kt',
+      '**/submission/TransmissionRiskVectorDeterminator.kt'
```
Comment (d4rken) [inline Corona-Warn-App/config/detekt.yml:526]: Why is this necessary?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt [hunk 17]
```diff
@@ -0,0 +1,44 @@
+package de.rki.coronawarnapp.submission
+
+import KeyExportFormat
+import com.google.android.gms.nearby.exposurenotification.TemporaryExposureKey
+import com.google.protobuf.ByteString
+
+class ExposureKeyDomain {
+
+    companion object {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt:8]: Move the compation object to the end of the class (kotlin style guide).

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt [hunk 18]
```diff
@@ -0,0 +1,44 @@
+package de.rki.coronawarnapp.submission
+
+import KeyExportFormat
+import com.google.android.gms.nearby.exposurenotification.TemporaryExposureKey
+import com.google.protobuf.ByteString
+
+class ExposureKeyDomain {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt:6]: Does `Domain` fit here? Isn't this used like a `ExposureKeyConverter`?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+    companion object {
+
+        private const val ROLLING_PERIOD = 144
+        private const val MAXIMUM_KEYS = 14
+    }
+
+    fun transformToKeyHistoryInExternalFormat(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt:13]: What about moving these to the constructor, so we have a 
`ExposureKeyConverter(arg1,arg2).createExternalFormat() : ReturnType`

If this is the "external api" for this classes logic, please add a specific return type, which makes it easier to understand  when viewing without IDE, i.e. currently I cant'see what the method should return.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/Symptoms.kt [hunk 20]
```diff
@@ -0,0 +1,6 @@
+package de.rki.coronawarnapp.submission
+
+class Symptoms(
+    val startOfSymptoms: StartOfSymptoms,
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/Symptoms.kt]: What about making `StartOfSymptoms` and `SymptomIndication` subclasses to keep it as one coherent unit (they are not used independently), i.e. `Symptoms.StartOf` and `Symptoms.Indication`. Not a *mustHave*  but seems cleaner to me.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/TransmissionRiskVector.kt [hunk 21]
```diff
@@ -0,0 +1,12 @@
+package de.rki.coronawarnapp.submission
+
+class TransmissionRiskVector(private val values: IntArray) {
+
+    companion object {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/TransmissionRiskVector.kt:5]: Bottom of the class please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculation.kt [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+import org.joda.time.Instant
+import org.joda.time.LocalDate
+import java.util.Locale
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class CalendarCalculation @Inject constructor() {
+
+    companion object {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculation.kt:13]: Bottom of the class please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculation.kt [hunk 23]
```diff
@@ -0,0 +1,120 @@
+package de.rki.coronawarnapp.ui.calendar
+
+import org.joda.time.DateTime
+import org.joda.time.Instant
+import org.joda.time.LocalDate
+import java.util.Locale
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculation.kt:10]: As this class is stateless, and not necessarily someone we want to keep in the graph for longer, mark this `@Reusable`, it will be reused, but not kept indefintely.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt [hunk 24]
```diff
@@ -50,24 +60,38 @@ class SubmissionViewModel : ViewModel() {
...
-        } catch (err: TransactionException) {
-            if (err.cause is CwaWebException) {
-                _submissionError.value = Event(err.cause)
-            } else {
-                err.report(ExceptionCategory.INTERNAL)
+    val symptomIndication = MutableLiveData<SymptomIndication?>()
+    val symptomStart = MutableLiveData<StartOfSymptoms?>()
+
+    fun initSymptoms() {
+        symptomIndication.postValue(null)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt:67]: Can't we just initialize `MutableLiveData<SymptomIndication?>(null)` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt [hunk 25]
```diff
@@ -50,24 +60,38 @@ class SubmissionViewModel : ViewModel() {
...
+    }
+    fun initSymptomStart() {
+        symptomStart.postValue(null)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt:70]: Can't we just initialize `MutableLiveData<StartOfSymptoms?>(null)` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt [hunk 26]
```diff
@@ -142,4 +166,52 @@ class SubmissionViewModel : ViewModel() {
             }
         }
     }
+
+    fun onNextClicked() {
+        symptomIntroductionEvent.value = SymptomIntroductionEvent.NavigateToSymptomCalendar
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt:171]: Is there a special requirement that you directly assign the value, and not use the async `postValue` mechanism here?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt [hunk 27]
```diff
@@ -142,4 +166,52 @@ class SubmissionViewModel : ViewModel() {
...
+    }
+
+    fun onPreviousClicked() {
+        symptomIntroductionEvent.value = SymptomIntroductionEvent.NavigateToPreviousScreen
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt:175]: `postValue`?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt [hunk 28]
```diff
@@ -142,4 +166,52 @@ class SubmissionViewModel : ViewModel() {
...
+    }
+
+    fun onCalendarNextClicked() {
+        symptomCalendarEvent.value = SymptomCalendarEvent.NavigateToNext
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt:179]: `postValue`?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt [hunk 29]
```diff
@@ -142,4 +166,52 @@ class SubmissionViewModel : ViewModel() {
...
+    }
+
+    fun onCalendarPreviousClicked() {
+        symptomCalendarEvent.value = SymptomCalendarEvent.NavigateToPrevious
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt:183]: `postValue`?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/Dates.kt [hunk 30]
```diff
@@ -0,0 +1,7 @@
+package de.rki.coronawarnapp.util
+
+object Dates {
+    private const val MILLIS_PER_DAY = 24 * 3600 * 1000
+
+    fun numberOfDays(t0: Long, t1: Long) = ((t1 - t0) / MILLIS_PER_DAY).toInt()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/Dates.kt:6]: Please replace custom time logic with framework classes whereever possible, i.e. here something like `Duration(Instant.now(),Instant.now()).standardDays`

File: Corona-Warn-App/src/main/res/values/strings.xml [hunk 31]
```diff
@@ -980,6 +980,47 @@
...
     <string name="submission_done_illustration_description">"Everyone in the group is cheering because someone has shared the test result."</string>
 
+    <!-- Submission Symptoms -->
+    <!-- XHED: Page title for symptom screens -->
+    <string name="submission_symptom_title">""</string>
+    <!-- YTXT: headline text for initial symptom screen -->
+    <string name="submission_symptom_initial_headline">""</string>
+    <!-- YTXT: Bullet points for symptoms -->
+    <string-array name="submission_symptom_symptom_bullets">
+        <item>""</item>
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/values/strings.xml:990]: Is it normal that these are empty?

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculationTest.kt (hunk 12 above)
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculationTest.kt:20]: Please use kotest for assertions, i.e. `dates.first().date.dayOfMonth shouldBe 3`

**Review by d4rken — CHANGES_REQUESTED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/Dates.kt (hunk 30 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/Dates.kt:6]: Likely remove the `Dates` class and use framework classes where `Dates.numberOfDays` is currently used.

File: Corona-Warn-App/src/main/res/values/strings.xml (hunk 31 above)
Comment (Oliver-Zimmerman) [reply Corona-Warn-App/src/main/res/values/strings.xml:990]: This is a part of the translation process. Provide the german string values, and leave the default string values empty.

File: Corona-Warn-App/config/detekt.yml (hunk 15 above)
Comment (Oliver-Zimmerman) [reply Corona-Warn-App/config/detekt.yml:103]: @AlexanderAlferov

File: Corona-Warn-App/config/detekt.yml (hunk 16 above)
Comment (Oliver-Zimmerman) [reply Corona-Warn-App/config/detekt.yml:526]: @AlexanderAlferov

File: Corona-Warn-App/config/detekt.yml (hunk 15 above)
Comment (chris-cwa) [reply Corona-Warn-App/config/detekt.yml:103]: because otherwise we could have only 20 methods max

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculationTest.kt (hunk 12 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculationTest.kt:20]: why is that necessary?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt (hunk 18 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt:6]: no, this does class does not do converting

Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt:6]: it contains logic for exposure key history

File: Corona-Warn-App/config/detekt.yml (hunk 16 above)
Comment (chris-cwa) [reply Corona-Warn-App/config/detekt.yml:526]: because transmissionriskvector is allowed to have magic umbers

File: Corona-Warn-App/config/detekt.yml (hunk 15 above)
Comment (jakobmoellersap) [reply Corona-Warn-App/config/detekt.yml:103]: Correct, this is for readability and better understanding of code. Please try to extract code out of the class if you need more than 20 methods

File: Corona-Warn-App/config/detekt.yml (hunk 16 above)
Comment (jakobmoellersap) [reply Corona-Warn-App/config/detekt.yml:526]: For me okay here, since we have a LOT of magic Numbers otherwise that are just set numbers in the Vector.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt (hunk 18 above)
Comment (jakobmoellersap) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyDomain.kt:6]: A Domain does not contain logic imo. I agree though that this is not pure conversion. Logic is used for interpreting keys that come out of the API and correctly assign Risk Levels as well, so maybe we should decouple and/or rename accordingly to purpose

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculationTest.kt (hunk 12 above)
Comment (jakobmoellersap) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculationTest.kt]: Right now that is our only test that uses TestCase. I would argue for kotest or JUnit Assertions to keep in line with our other test suite components

File: Corona-Warn-App/config/detekt.yml (hunk 15 above)
Comment (d4rken) [reply Corona-Warn-App/config/detekt.yml:103]: I agree that incase of ViewModels it would make code less readable to abstract it just due to method count. Let's go forward with the exclusion but please use `**/*ViewModel.kt` (test this before).

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt (hunk 26 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt]: @Oliver-Zimmerman

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt (hunk 27 above)
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt]: @Oliver-Zimmerman

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/SingleLiveEvent.kt [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ *
+ * This avoids a common problem with events: on configuration change (like rotation) an update
+ * can be emitted if the observer is active. This LiveData only calls the observable if there's an
+ * explicit call to setValue() or call().
+ *
+ *
+ * Note that only one observer is going to be notified of changes.
+ */
+class SingleLiveEvent<T> : MutableLiveData<T>() {
```
[the remark this answers is not in the document]
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/SingleLiveEvent.kt:39]: @d4rken will pull and solve this to his pr

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculation.kt [hunk 33]
```diff
@@ -0,0 +1,120 @@
...
+class CalendarCalculation @Inject constructor() {
```
[the remark this answers is not in the document]
Comment (chris-cwa) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/calendar/CalendarCalculation.kt:12]: added new year and leap year test

**Review by jakobmoellerdev — DISMISSED** [review]

**Review by d4rken — DISMISSED** [review]
lgtm

**Review by harambasicluka — DISMISSED** [review]
